File: carelog/src/services/fileService.ts
import * as FileSystem from 'expo-file-system';
import * as ImageManipulator from 'expo-image-manipulator';
import { AttachmentType } from '@src/types/Attachment';

const BASE_DIR = `${FileSystem.documentDirectory}carelog/`;
const ATTACHMENTS_DIR = `${BASE_DIR}attachments/`;

function visitDir(visitId: string): string {
  return `${ATTACHMENTS_DIR}${visitId}/`;
}

async function ensureDir(path: string): Promise<void> {
  await FileSystem.makeDirectoryAsync(path, { intermediates: true });
}

function extForMime(mimeType: string): string {
  if (mimeType === 'application/pdf') return '.pdf';
  if (mimeType === 'image/png') return '.png';
  return '.jpg';
}

export const fileService = {
  /**
   * Copies sourceUri into carelog/attachments/{visitId}/, optionally compressing
   * and generating a thumbnail for images.
   */
  async saveAttachment(
    visitId: string,
    type: AttachmentType,
    sourceUri: string,
    mimeType: string,
  ): Promise<{ filePath: string; fileName: string; thumbnailPath?: string; sizeBytes: number }> {
    const dir = visitDir(visitId);
    await ensureDir(dir);

    const ext = extForMime(mimeType);
    const fileName = `${type}_${Date.now()}${ext}`;
    const isImage = mimeType.startsWith('image/');

    // Compress before saving if it's an image
    const processedUri = isImage
      ? await fileService.compressImage(sourceUri)
      : sourceUri;

    const filePath = `${dir}${fileName}`;
    await FileSystem.copyAsync({ from: processedUri, to: filePath });

    const info = await FileSystem.getInfoAsync(filePath);
    const sizeBytes = info.exists && !info.isDirectory ? (info as any).size ?? 0 : 0;

    let thumbnailPath: string | undefined;
    if (isImage) {
      thumbnailPath = await fileService.generateThumbnail(filePath);
    }

    return { filePath, fileName, thumbnailPath, sizeBytes };
  },

  /** Deletes a file (and optionally its thumbnail) from the filesystem. */
  async deleteAttachment(filePath: string): Promise<void> {
    const info = await FileSystem.getInfoAsync(filePath);
    if (info.exists) await FileSystem.deleteAsync(filePath, { idempotent: true });
  },

  /**
   * Resizes an image to fit within 200×200 and saves it to the thumbnails
   * subdirectory. Returns the path of the saved thumbnail.
   */
  async generateThumbnail(filePath: string): Promise<string> {
    const thumbDir = `${BASE_DIR}thumbnails/`;
    await ensureDir(thumbDir);

    const result = await ImageManipulator.manipulateAsync(
      filePath,
      [{ resize: { width: 200, height: 200 } }],
      { compress: 0.8, format: ImageManipulator.SaveFormat.JPEG },
    );

    const thumbPath = `${thumbDir}thumb_${Date.now()}.jpg`;
    await FileSystem.copyAsync({ from: result.uri, to: thumbPath });
    return thumbPath;
  },

  /**
   * Compresses an image to max 2048px on the longest edge at quality 85.
   * Returns the URI of the compressed image (in the system tmp cache).
   */
  async compressImage(uri: string): Promise<string> {
    // First pass with no actions to read dimensions
    const probe = await ImageManipulator.manipulateAsync(uri, [], {});
    const { width, height } = probe;
    const longestEdge = Math.max(width, height);

    const actions: ImageManipulator.Action[] = [];
    if (longestEdge > 2048) {
      const scale = 2048 / longestEdge;
      actions.push({
        resize: {
          width: Math.round(width * scale),
          height: Math.round(height * scale),
        },
      });
    }

    const result = await ImageManipulator.manipulateAsync(
      uri,
      actions,
      { compress: 0.85, format: ImageManipulator.SaveFormat.JPEG },
    );
    return result.uri;
  },

  /** Sums the size of every file under carelog/attachments/. */
  async getStorageUsedBytes(): Promise<number> {
    const info = await FileSystem.getInfoAsync(ATTACHMENTS_DIR);
    if (!info.exists) return 0;

    let total = 0;
    const visitFolders = await FileSystem.readDirectoryAsync(ATTACHMENTS_DIR);
    for (const folder of visitFolders) {
      const folderPath = `${ATTACHMENTS_DIR}${folder}/`;
      const folderInfo = await FileSystem.getInfoAsync(folderPath);
      if (!folderInfo.exists || !folderInfo.isDirectory) continue;
      const files = await FileSystem.readDirectoryAsync(folderPath);
      for (const file of files) {
        const fInfo = await FileSystem.getInfoAsync(`${folderPath}${file}`);
        if (fInfo.exists && !fInfo.isDirectory) total += (fInfo as any).size ?? 0;
      }
    }
    return total;
  },

  /** Deletes the entire carelog/ directory (all attachments and thumbnails). */
  async deleteAllAttachments(): Promise<void> {
    const info = await FileSystem.getInfoAsync(BASE_DIR);
    if (info.exists) await FileSystem.deleteAsync(BASE_DIR, { idempotent: true });
  },
};

File: flutter_version/lib/services/file_service.dart
import 'dart:io';

import 'package:image/image.dart' as img;
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:uuid/uuid.dart';

class SavedFile {
  final String filePath;
  final String fileName;
  final String? thumbnailPath;
  final int sizeBytes;

  const SavedFile({
    required this.filePath,
    required this.fileName,
    this.thumbnailPath,
    required this.sizeBytes,
  });
}

/// Manages on-disk storage for visit attachments and insurance documents.
/// Files are organised under `<appDocs>/attachments/<ownerId>/` where ownerId
/// is either a visit UUID or a policy UUID — they never collide.
class FileService {
  static const _uuid = Uuid();

  Future<String> _attachmentsRoot() async {
    final docs = await getApplicationDocumentsDirectory();
    return p.join(docs.path, 'attachments');
  }

  String _extForMime(String mimeType) {
    if (mimeType == 'application/pdf') return '.pdf';
    if (mimeType == 'image/png') return '.png';
    return '.jpg';
  }

  /// Copies [sourcePath] into `attachments/<ownerId>/`, compressing images and
  /// generating a thumbnail (max 400 px on the longest edge).
  Future<SavedFile> saveFile(
    String ownerId,
    String label,
    String sourcePath,
    String mimeType,
  ) async {
    final root = await _attachmentsRoot();
    final dir = p.join(root, ownerId);
    await Directory(dir).create(recursive: true);

    final ext = _extForMime(mimeType);
    final stamp = DateTime.now().millisecondsSinceEpoch;
    final fileName = '${label}_$stamp$ext';
    final destPath = p.join(dir, fileName);
    final isImage = mimeType.startsWith('image/');

    String? thumbnailPath;

    if (isImage) {
      final bytes = await File(sourcePath).readAsBytes();
      final decoded = img.decodeImage(bytes);
      if (decoded != null) {
        // Compress: limit longest edge to 2048 px
        img.Image processed = decoded;
        final longest = decoded.width > decoded.height ? decoded.width : decoded.height;
        if (longest > 2048) {
          final scale = 2048 / longest;
          processed = img.copyResize(
            decoded,
            width: (decoded.width * scale).round(),
            height: (decoded.height * scale).round(),
          );
        }
        await File(destPath).writeAsBytes(img.encodeJpg(processed, quality: 85));

        // Thumbnail: limit longest edge to 400 px
        final tLongest = processed.width > processed.height ? processed.width : processed.height;
        final tScale = 400 / tLongest;
        final thumb = img.copyResize(
          processed,
          width: (processed.width * tScale).round(),
          height: (processed.height * tScale).round(),
        );
        final thumbName = 'thumb_${_uuid.v4()}.jpg';
        final thumbPath = p.join(dir, thumbName);
        await File(thumbPath).writeAsBytes(img.encodeJpg(thumb, quality: 80));
        thumbnailPath = thumbPath;
      } else {
        // Unrecognised image format — copy as-is
        await File(sourcePath).copy(destPath);
      }
    } else {
      await File(sourcePath).copy(destPath);
    }

    final sizeBytes = await File(destPath).length();
    return SavedFile(
      filePath: destPath,
      fileName: fileName,
      thumbnailPath: thumbnailPath,
      sizeBytes: sizeBytes,
    );
  }

  Future<SavedFile> saveAttachment(
    String visitId,
    String type,
    String sourcePath,
    String mimeType,
  ) =>
      saveFile(visitId, type, sourcePath, mimeType);

  Future<SavedFile> saveInsuranceDocument(
    String policyId,
    String sourcePath,
    String mimeType,
  ) =>
      saveFile(policyId, 'insurance', sourcePath, mimeType);

  Future<void> deleteFiles(List<String> paths) async {
    for (final path in paths) {
      final file = File(path);
      if (await file.exists()) await file.delete();
    }
  }

  Future<void> deleteOwnerDir(String ownerId) async {
    final root = await _attachmentsRoot();
    final dir = Directory(p.join(root, ownerId));
    if (await dir.exists()) await dir.delete(recursive: true);
  }

  Future<void> deleteAllAttachments() async {
    final root = await _attachmentsRoot();
    final dir = Directory(root);
    if (await dir.exists()) await dir.delete(recursive: true);
  }

  /// Sums the size of every file under the attachments root.
  Future<int> getStorageUsedBytes() async {
    final root = await _attachmentsRoot();
    final dir = Directory(root);
    if (!await dir.exists()) return 0;
    int total = 0;
    await for (final entity in dir.list(recursive: true)) {
      if (entity is File) total += await entity.length();
    }
    return total;
  }
}

final fileService = FileService();
